src/ApacheOrcDotNet.OptimizedReader/OrcReader.cs

```diff
@@ -35,93 +40,452 @@ public OrcReader(OrcReaderConfiguration configuration, IByteRangeProvider byteRa
-            var rowIndexStream = streamsForColumn.First(s => s.StreamKind == StreamKind.RowIndex);
+        public IEnumerable<int> GetRowGroupIndexes(int stripeId, OrcColumn column, string min, string max)
+        {
+            var rowIndex = GetColumnRowIndex(column.Id, stripeId);
+            return GetRowGroupIndexes(Enumerable.Range(0, rowIndex.Entry.Count), stripeId, column, min, max);
+        }
 
-            var result = _byteRangeProvider.DecompressAndParseByteRange(
-                rowIndexStream.FileOffset,
-                rowIndexStream.Length,
-                _fileTail.PostScript.Compression,
-                (int)_fileTail.PostScript.CompressionBlockSize,
-                sequence => SpanRowGroupIndex.ReadRowGroupDetails(sequence, streamsForColumn, _fileTail.PostScript.Compression)
-            );
+        public IEnumerable<int> GetRowGroupIndexes(IEnumerable<int> lookupIndexes, int stripeId, OrcColumn column, string min, string max)
+        {
+            var rowIndex = GetColumnRowIndex(column.Id, stripeId);
 
-            return result;
+            return lookupIndexes.Where(index =>
+            {
+                var rowIndexEntry = rowIndex.Entry[index];
+                return rowIndexEntry.Statistics.InRange(column, min, max);
+            });
         }
 
-        public int ReadLongValues(Span<long> output, StreamPosition position)
+        public async Task LoadDataAsync<TOutput>(int stripeId, int rowEntryIndexId, BaseColumnBuffer<TOutput> columnBuffer)
         {
-            throw new NotImplementedException();
+            var rowIndex = GetColumnRowIndex(columnBuffer.Column.Id, stripeId);
+            var columnStreams = GetColumnDataStreams(stripeId, columnBuffer.Column, rowIndex, rowEntryIndexId);
+
+            await columnBuffer.LoadDataAsync(stripeId, columnStreams);
         }
 
-        IEnumerable<StreamDetail> ReadStripeFooter(int stripeId)
+        public void Fill<TOutput>(BaseColumnBuffer<TOutput> columnBuffer, bool discardPreviousData = true)
         {
-            var stripe = _fileTail.Footer.Stripes[stripeId];
-            var stripeFooterStart = (int)(stripe.Offset + stripe.IndexLength + stripe.DataLength); //TODO consider supporting >2TB files here
-            var stripeFooterLength = (int)stripe.FooterLength;
+            if (discardPreviousData)
+                columnBuffer.Reset();
 
-            var result = _byteRangeProvider.DecompressAndParseByteRange(
-                stripeFooterStart,
-                stripeFooterLength,
-                _fileTail.PostScript.Compression,
-                (int)_fileTail.PostScript.CompressionBlockSize,
-                sequence => SpanStripeFooter.ReadStreamDetails(sequence, ColumnDetails, (long)stripe.Offset)
-            );
+            columnBuffer.Fill();
 
-            return result;
+            if (NumValues == 0 || NumValues > columnBuffer.Values.Length)
+                NumValues = columnBuffer.Values.Length;
```

Comment:
This doesn't seem threadsafe.  It looks like you're trying to keep this class threadsafe, based on your use of `ConcurrentDictionary`.

Reply:
This method has been updated and we are accessing the `columnBuffer` only after it's already filled with values.